libbpf-tools/cachestat.c

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: (LGPL-2.1 OR BSD-2-Clause)
+// Copyright (c) 2021 Wenbo Zhang
+//
+// Based on cachestat(8) from BCC by Brendan Gregg and Allan McAleavy.
+// 8-Mar-2021   Wenbo Zhang   Created this.
+#include <argp.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <time.h>
+#include <bpf/libbpf.h>
+#include <bpf/bpf.h>
+#include "cachestat.h"
+#include "cachestat.skel.h"
+#include "trace_helpers.h"
+
+static struct env {
+	time_t interval;
+	int times;
+	bool timestamp;
+	bool verbose;
+} env = {
+	.interval = 1,
+	.times = 99999999,
+};
+
+static volatile bool exiting;
+
+const char *argp_program_version = "cachestat 0.1";
+const char *argp_program_bug_address = "<bpf@vger.kernel.org>";
+const char argp_program_doc[] =
+"Count cache kernel function calls.\n"
+"\n"
+"USAGE: cachestat [--help] [-T] [interval] [count]\n"
+"\n"
+"EXAMPLES:\n"
+"    cachestat           # shows hits and misses to the file system page cache"
+"    cachestat -T        # include timestamps"
+"    cachestat 1 10      # print 1 second summaries, 10 times\n";
+
+static const struct argp_option opts[] = {
+	{ "timestamp", 'T', NULL, 0, "Print timestamp" },
+	{ "verbose", 'v', NULL, 0, "Verbose debug output" },
+	{},
+};
+
+static error_t parse_arg(int key, char *arg, struct argp_state *state)
+{
+	static int pos_args;
+
+	switch (key) {
+	case 'v':
+		env.verbose = true;
+		break;
+	case 'T':
+		env.timestamp = true;
+		break;
+	case ARGP_KEY_ARG:
+		errno = 0;
+		if (pos_args == 0) {
+			env.interval = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid internal\n");
+				argp_usage(state);
+			}
+		} else if (pos_args == 1) {
+			env.times = strtol(arg, NULL, 10);
+			if (errno) {
+				fprintf(stderr, "invalid times\n");
+				argp_usage(state);
+			}
+		} else {
+			fprintf(stderr,
+				"unrecognized positional argument: %s\n", arg);
+			argp_usage(state);
+		}
+		pos_args++;
+		break;
+	default:
+		return ARGP_ERR_UNKNOWN;
+	}
+	return 0;
+}
+
+int libbpf_print_fn(enum libbpf_print_level level,
+		    const char *format, va_list args)
+{
+	if (level == LIBBPF_DEBUG && !env.verbose)
+		return 0;
+	return vfprintf(stderr, format, args);
+}
+
+static void sig_handler(int sig)
+{
+	exiting = true;
+}
+
+static int get_meminfo(__u64 *buffers, __u64 *cached)
+{
+	__u64 tmp;
+	FILE *f;
+
+	f = fopen("/proc/meminfo", "r");
+	if (!f)
+		return -1;
+	if (fscanf(f, "MemTotal: %llu kB MemFree: %llu kB MemAvailable: %llu "
```

Comment:
nit: for a bit better readability, why don't you format the format string with one item a line:

```
"MemTotal: %llu kB\n"
"MemFree: %llu kB\n"
...
```

Also note that you can do "%*llu" to ignore that format (you won't need `tmp` then).

Reply:
Fixed, thanks!